compiler/src/dotty/tools/dotc/transform/Erasure.scala

```diff
@@ -550,15 +550,41 @@ object Erasure {
       if (!ctx.mode.is(Mode.Type)) {
         if (isErased(tree))
           report.error(em"${tree.symbol} is declared as erased, but is in fact used", tree.srcPos)
-        tree.symbol.getAnnotation(defn.CompileTimeOnlyAnnot) match {
-          case Some(annot) =>
-            def defaultMsg =
-              i"""Reference to ${tree.symbol.showLocated} should not have survived,
-                 |it should have been processed and eliminated during expansion of an enclosing macro or term erasure."""
-            val message = annot.argumentConstant(0).fold(defaultMsg)(_.stringValue)
-            report.error(message, tree.srcPos)
-          case _ => // OK
-        }
+
+        def parentSyms(tpe: Type): LazyList[Symbol] =
+          val ref = tpe.widen.finalResultType
+          ref.classSymbol #:: ref.parents.to(LazyList).flatMap(parentSyms)
+
+        def checkTree(tree: Tree, pos: util.SrcPos, toCheck: LazyList[Symbol]): Boolean =
+          toCheck.exists { sym =>
+            sym.getAnnotation(defn.CompileTimeOnlyAnnot) match {
+              case Some(annot) =>
+                def defaultMsg =
+                  i"""Reference to ${tree.symbol.showLocated} should not have survived,
+                    |it should have been processed and eliminated during expansion of an enclosing macro or term erasure."""
+                val message = annot.argumentConstant(0).fold(defaultMsg)(_.stringValue)
+                report.error(message, pos)
+                true
+              case _ => // OK
+                false
+            }
+          }
+
+        tree match
+          case ddef: DefDef =>
+            for
+              vparams <- ddef.vparamss
```

Comment:
RefChecks already traverses trees, can we check things at the point where we're traversing them instead of retraversing them here? This is what scala 2 does (see `checkUndesiredProperties`)

Reply:
following discussion at the dotty meeting this will need to be done at the same place we issue deprecation warnings, as scalac reports violations for type aliases.

Possibly follow what is described in https://github.com/lampepfl/dotty/issues/9825#issuecomment-720601904